javascript/ql/src/Performance/ReDoS.ql

```diff
@@ -873,7 +878,7 @@ from RegExpTerm t, string c, int i
 where
   c = min(string w | isPumpable(Match(t, i), w)) and
   not isPumpable(epsilonSucc+(Match(t, i)), _) and
-  not epsilonSucc*(process(Match(t, i), c, c.length() - 1)) = Accept(_)
+  not epsilonSucc*(process(Match(t, i), c, [0 .. c.length() - 1])) = Accept(_)
```

Comment:
> If we change [0 .. c.length() - 1] back to c.length() - 1, then the analysis concludes that it can have exponential backtracking with the witness string AXA.

Yes, but I didn't think this check was the root cause of that FP. I saw the check and thought it was suspicious that it didn't match the theory laid out at the top of the query. It wasn't obvious _why_ passing through an accept state mattered here. In such cases, it often pays off to get to the bottom of it, even if it's tempting to just use what appears to work and move on.

After giving it some more thought, I believe it has to do with how we model anchors (or lack of anchors). There's an ambiguity around what the accept state means. Because of our loose handling of anchors, being in the accept state could mean one of two things:
1. I accept if the string ends here (the text book interpretation of the accept state), or
2. I can just stop matching here if I want (the unanchored regexp interpretation).

The old check made sense assuming (1), but given how we map JS regexps to an NFA, (2) is closer to reality.

If we were to introduce proper handling of anchoring, I think the old check would need to be reinstated, but with the current NFA construction, we need to emulate the (2) interpretation. I see that the modified check here was a way to do that, and I'm happy if you want to leave it at that (but please explain in a comment). A more direct way to do the same might be to add an `Any()` transition from the accept state to itself.

----

As an aside, we could also try to implement better anchor support. That might be well out of scope for this PR, but just a few thoughts:

We could introduce a new state `MkAcceptAnySuffix()` which accepts an arbitrary suffix from that point on. It can be modelled via an epsilon transition to `MkAccept()` and an `Any()` edge to itself.

With that state in hand, we have a way to model anchors:
- At a `$` anchor, add transition to `MkAccept()` (which only succeeds if the string ends here and now)
- At the end of the regexp, transition to `MkAcceptAnySuffix()`.

We might lose some TPs though, since only testing the empty string probably won't be enough anymore. Maybe the empty string check only works so well because $-anchored regexps have an unreachable accept state.

Reply:
Great writeup :heart: 

Yes, I had to some extent overlooked how the accept state works compared to the theory.  

> I see that the modified check here was a way to do that, and I'm happy if you want to leave it at that (but please explain in a comment)   

I think we should keep the implementation as is for now, and I'll put in a comment at the `process()` predicate about which assumptions is makes about the accept state.   

I'll look into anchors again sometime after merging this PR. 

Note: I'm somewhat sure we don't need to model the start anchor `^`.